components/localize/localize-mixin.js

```diff
@@ -0,0 +1,158 @@
+import IntlMessageFormat from 'intl-messageformat/src/main.js';
+window.IntlMessageFormat = IntlMessageFormat;
+
+export const LocalizeMixin = superclass => class extends superclass {
+
+	static get properties() {
+		return {
+			language: {type: String},
+			langResources: { type: Object },
+			__documentLanguage: { type: String },
+			__documentLanguageFallback: { type: String }
+		};
+	}
+
+	connectedCallback() {
+		if (super.connectedCallback) {
+			super.connectedCallback();
+		}
+	}
+
+	constructor() {
+		super();
+		this.__documentLanguage = window.document.getElementsByTagName('html')[0].getAttribute('lang');
+		this.__documentLanguageFallback = window.document.getElementsByTagName('html')[0].getAttribute('data-lang-default');
+		this._startObserver();
+	}
+
+	firstUpdated(changedProperties) {
+		changedProperties.forEach((oldValue, propName) => {
+			if (propName === 'langResources') {
+				this._computeLanguage();
+			}
+		});
+	}
+
+	updated(changedProperties) {
+		changedProperties.forEach((oldValue, propName) => {
+			if (propName === 'language') {
+				this._languageChange();
+			}
+			// to do: add __timezoneObject which calls _timezoneChange()
+		});
+	}
+
+	localize(key) {
+		return this._computeLocalize(this.language, this.langResources, key);
+	}
+
+	_startObserver() {
+		var htmlElem = window.document.getElementsByTagName('html')[0];
+
+		this._observer = new MutationObserver(function(mutations) {
+			for (var i = 0; i < mutations.length; i++) {
+				var mutation = mutations[i];
+				if (mutation.attributeName === 'lang') {
+					this.__documentLanguage = htmlElem.getAttribute('lang');
+				} else if (mutation.attributeName === 'data-lang-default') {
+					this.__documentLanguageFallback = htmlElem.getAttribute('data-lang-default');
+				} else if (mutation.attributeName === 'data-intl-overrides') {
+					this.__overrides = this._tryParseHtmlElemAttr('data-intl-overrides', {});
+				} else if (mutation.attributeName === 'data-timezone') {
+					this.__timezoneObject = this._tryParseHtmlElemAttr('data-timezone', {name: '', identifier: ''});
+				}
+			}
+		}.bind(this));
+		this._observer.observe(htmlElem, { attributes: true });
+	}
+
+	_computeLanguage() {
+		this.language = this._tryResolve(this.langResources, this.__documentLanguage)
+			|| this._tryResolve(this.langResources, this.__documentLanguageFallback)
+			|| this._tryResolve(this.langResources, 'en-us');
+	}
+
+	_tryResolve(resources, val) {
+
+		if (val === null) return null;
+		val = val.toLowerCase();
+		var baseLang = val.split('-')[0];
+
+		var foundBaseLang = null;
+		for (var key in resources) {
+			var keyLower = key.toLowerCase();
+			if (keyLower.toLowerCase() === val) {
+				return key;
+			} else if (keyLower === baseLang) {
+				foundBaseLang = key;
+			}
+		}
+
+		if (foundBaseLang) {
+			return foundBaseLang;
+		}
+
+		return null;
+	}
+
+	_languageChange() {
+		this.fire('d2l-localize-behavior-language-changed');
```

Comment:
I think we should use the dispatch syntax instead of fire.

Reply:
It's unfortunate that `behavior` was part of the event name.  I'd consider changing depending on what listens to this.

Reply:
I _think_ this was only used in tests/demo? We could check and rename.

Reply:
it's used in a few places that aren't tests/demo https://github.com/search?q=org%3ABrightspace+d2l-localize-behavior-language-changed&type=Code

Reply:
Hmm, unfortunate. If we wanted to rename, we could just fire both events I guess.